File: schemas/notifications-system/convex/email_suppression_list.ts
// email_suppression_list: Email addresses that should not be sent to. Prevents bounces, spam complaints, and unsubscribe violations.
// See README.md for full design rationale.

import { defineTable } from "convex/server";
import { v } from "convex/values";

export const email_suppression_list = defineTable({
  email: v.string(), // Lowercase, trimmed.

  // Why this address is suppressed.
  reason: v.union(
    v.literal("hard_bounce"), // Mailbox doesn't exist (permanent). Never send again.
    v.literal("soft_bounce"), // Temporary delivery failure. May clear after a cooling period.
    v.literal("spam_complaint"), // Recipient marked your email as spam. Never send again.
    v.literal("manual_unsubscribe"), // User clicked unsubscribe link. Respect immediately.
    v.literal("invalid_address") // Address format is invalid or domain doesn't exist.
  ),

  // How this suppression was created.
  source: v.union(
    v.literal("provider_webhook"), // Bounce/complaint webhook from SendGrid, Postmark, etc.
    v.literal("user_action"), // User clicked unsubscribe link in your app.
    v.literal("admin"), // Manually added by an admin.
    v.literal("system") // Automated detection (e.g., repeated soft bounces promoted to hard bounce).
  ),

  channelId: v.optional(v.id("notification_channels")), // Which provider reported the suppression.

  details: v.optional(v.any()), // Provider-specific details for debugging (e.g., bounce_type, provider_message).

  suppressedAt: v.number(), // When the suppression took effect. May differ from _creationTime if back-dated.
  expiresAt: v.optional(v.number()), // Null = permanent suppression. Set for soft bounces that should be retried.
  // no createdAt — Convex provides _creationTime
})
  .index("by_email_reason", ["email", "reason"]) // unique(email, reason): one entry per email per reason.
  .index("by_email", ["email"]) // "Is this email suppressed?" (checked before every email delivery).
  .index("by_reason", ["reason"]) // "All hard bounces" (for reporting).
  .index("by_expires_at", ["expiresAt"]); // Cleanup job: remove expired soft bounce suppressions.

File: schemas/notifications-system/spacetimedb/email_suppression_list.rs
// email_suppression_list: Email addresses that should not be sent to. Prevents bounces, spam complaints, and unwanted mail.
// See README.md for full design rationale.

use spacetimedb::{SpacetimeType, Timestamp};

// Why this address is suppressed.
// type: String
#[derive(SpacetimeType, Clone)]
pub enum SuppressionReason {
    HardBounce,       // Mailbox doesn't exist (permanent). Never send again.
    SoftBounce,       // Temporary delivery failure. May clear after a cooling period.
    SpamComplaint,    // Recipient marked your email as spam. Never send again.
    ManualUnsubscribe, // User clicked unsubscribe link. Respect immediately.
    InvalidAddress,   // Address format is invalid or domain doesn't exist.
}

// How this suppression was created.
// type: String
#[derive(SpacetimeType, Clone)]
pub enum SuppressionSource {
    ProviderWebhook, // Bounce/complaint webhook from SendGrid, Postmark, etc.
    UserAction,      // User clicked unsubscribe link in your app.
    Admin,           // Manually added by an admin.
    System,          // Automated detection (e.g., repeated soft bounces promoted to hard bounce).
}

#[spacetimedb::table(name = email_suppression_list, public)]
pub struct EmailSuppressionEntry {
    #[primary_key]
    pub id: String, // UUID

    #[index(btree)]
    pub email: String, // The suppressed email address. Lowercase, trimmed.

    #[index(btree)]
    pub reason: SuppressionReason,

    pub source: SuppressionSource,

    pub channel_id: Option<String>, // UUID — FK → notification_channels.id (set null on delete)

    pub details: Option<String>, // JSON blob: bounce_type, provider_message, original_message_id, etc.

    pub suppressed_at: Timestamp, // When the suppression took effect. May differ from created_at if back-dated.

    #[index(btree)]
    pub expires_at: Option<Timestamp>, // Null = permanent suppression. Set for soft bounces that should be retried.

    pub created_at: Timestamp,

    // Unique constraint: (email, reason) — one entry per email per reason.
    // SpacetimeDB does not enforce composite unique constraints; enforce in application logic.
}
